### PR TITLE
Register target for roles

### PR DESCRIPTION
**Pull request by neihousaigaai**

### CODE REVIEW

File: game/roles/hunter.py [hunk 1]
```diff
@@ -1,19 +1,32 @@
...
 
-    def set_hunted_target(self, hunted):
-        self.hunted_target = hunted
+    def set_target(self, target_id):
+        self.target = target_id
+
+    def is_valid_target(self, target_id):
+        return target_id is not None and target_id != self.player_id
+
+    def register_target(self, target):
```
Comment (nitro2) [inline game/roles/hunter.py:25]: Move register_target to Character class. 
Eg: character.py:
```python
def register_target(self, target, after_action_text):
        target_id = target.player_id
        if self.is_valid_target(target_id):
            self.set_target(target_id)
        else:
            self.set_target(None)

        return text_templates.generate_text(after_action_text, target=f"<@{target_id}>")
```
In this hunter.py, we have: 
```python
def register_target(self, target)
        return super().register_target(target, "hunter_after_voting_text")
```

File: game/roles/seer.py [hunk 2]
```diff
@@ -21,3 +22,12 @@ def get_target(self):
 
     def set_target(self, target_id):
         self.target = target_id
+
+    def register_target(self, target):
+        if self.get_mana() == 0:
+            return text_templates.generate_text("out_of_mana_text")
```
Comment (nitro2) [inline game/roles/seer.py:28]: Same as Guard

File: game/roles/witch.py [hunk 3]
```diff
@@ -41,3 +42,21 @@ def get_curse_target(self):
...
         self.curse_target = target_id
+
+    def register_reborn_target(self, target):
+        if self.get_power() == 0:
+            return text_templates.generate_text("out_of_power_text")
+
+        target_id = target.player_id
+        self.set_reborn_target(target_id)
+
+        return text_templates.generate_text("witch_after_reborn_text", target=f"<@{target_id}>")
```
Comment (nitro2) [inline game/roles/witch.py:53]:
```suggestion
    def register_reborn_target(self, target):
        self.set_reborn_target(target.player_id)
        return super().register_target(target, "witch_after_reborn_text")
```

File: game/roles/witch.py [hunk 4]
```diff
@@ -41,3 +42,21 @@ def get_curse_target(self):
...
+
+    def register_curse_target(self, target):
+        if self.get_curse_power() == 0:
+            return text_templates.generate_text("out_of_power_text")
+
+        target_id = target.player_id
+        self.set_curse_target(target_id)
+
+        return text_templates.generate_text("witch_after_curse_text", target=f"<@{target_id}>")
```
Comment (nitro2) [inline game/roles/witch.py:62]: Similar to reborn

**Review by nitro2 — COMMENTED** [review]
If we do not use any field other than target.player_id, should we directly use target_id for register_target()?

File: .pylintrc [hunk 5]
```diff
@@ -14,5 +14,6 @@ disable=
     W0105, # pointless-string-statement
     W0201, # attribute-defined-outside-init
     W0612, # unused-variable
+    R0801, # duplicate-code
```
[the remark this answers is not in the document]
Comment (neihousaigaai) [reply .pylintrc:17]: It occurs in class Seer and class Guard. 
```python
    def get_target(self):
        return self.target

    def set_target(self, target_id):
        self.target = target_id

    def register_target(self, target):
        if self.get_mana() == 0:
            return text_templates.generate_text("out_of_mana_text")

        target_id = target.player_id
        self.set_target(target_id)
```

[the remark this answers is not in the document]
Comment (nitro2) [reply .pylintrc:17]: I think we can solve it by moving these functions to Character class

File: game/__init__.py [hunk 6]
```diff
@@ -1252,24 +1206,15 @@ async def ship(self, author, target1, target2):
...
+        return author.register_target(target)
 
-    # Kill anyone who is hunted
     async def get_hunted_target_on_hunter_death(self, hunter):
+        """Kill anyone who is hunted"""
         if isinstance(self.players[hunter], roles.Hunter):
-            hunted = self.players[hunter].get_hunted_target()
-            if hunted and hunted != hunter:
+            hunted = self.players[hunter].get_target()
+            if self.players[hunter].is_valid_target(hunted):
```
[the remark this answers is not in the document]
Comment (neihousaigaai) [reply game/__init__.py:1217]: `is_valid_target(target)` also check `target is not None` first. I feel a bit confused.

File: game/roles/guard.py [hunk 7]
```diff
@@ -28,3 +29,18 @@ def get_target(self):
 
     def set_target(self, target_id):
         self.target = target_id
+
+    def register_target(self, target, allow_guard_self_protection):
+        if self.get_mana() == 0:
+            return text_templates.generate_text("out_of_mana_text")
```
[the remark this answers is not in the document]
Comment (neihousaigaai) [reply game/roles/guard.py:35]: `mana` property hasn't been used, will remove it from Character class.

File: game/__init__.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (nitro2) [reply game/__init__.py:1217]: Sorry, I was confused also. My previous statement is incorrect.
Why did you change the code? Why don't we use `if hunted and hunted != hunter:` ?

[the remark this answers is not in the document]
Comment (neihousaigaai) [reply game/__init__.py:1217]: I have moved `hunted and hunted != hunter` check to the new `is_valid_target` method. If user enters their id, I thought maybe we should assign `None` to remove their previous target.

I will revert to the old code.

**Review by nitro2 — APPROVED** [review]
Overview looks good. Other refactors need to address in new PR

**Pull request merged**